src/realm/query_engine.hpp

```diff
@@ -2470,35 +2475,37 @@ class NotNode : public ParentNode {
     size_t find_first_no_overlap(size_t start, size_t end);
 };
 
-
 // Compare two columns with eachother row-by-row
-template <class LeafType, class TConditionFunction>
-class TwoColumnsNode : public ParentNode {
+class TwoColumnsNodeBase : public ParentNode {
 public:
-    using TConditionValue = typename LeafType::value_type;
-
-    TwoColumnsNode(ColKey column1, ColKey column2)
+    TwoColumnsNodeBase(ColKey column1, ColKey column2)
     {
         m_dT = 100.0;
         m_condition_column_key1 = column1;
         m_condition_column_key2 = column2;
-    }
 
-    ~TwoColumnsNode() noexcept override
-    {
+        if (m_condition_column_key1.is_collection() || m_condition_column_key2.is_collection()) {
+            throw std::runtime_error(util::format("queries comparing two properties are not yet supported for "
+                                                  "collections (list/set/dictionary) ($1 and $2)",
+                                                  ParentNode::m_table->get_column_name(m_condition_column_key1),
+                                                  ParentNode::m_table->get_column_name(m_condition_column_key2)));
+        }
     }
 
+    ~TwoColumnsNodeBase() noexcept override {}
+
+    static std::unique_ptr<ArrayPayload> update_cached_leaf_pointers_for_column(Allocator& alloc,
+                                                                                const ColKey& col_key);
+    static Mixed get_value_from_leaf(ArrayPayload* leaf, ColumnType col_type, bool nullable, size_t ndx);
+
     void cluster_changed() override
     {
-        m_array_ptr1 = nullptr;
-        m_array_ptr1 = LeafPtr(new (&m_leaf_cache_storage1) LeafType(m_table.unchecked_ptr()->get_alloc()));
-        this->m_cluster->init_leaf(this->m_condition_column_key1, m_array_ptr1.get());
-        m_leaf_ptr1 = m_array_ptr1.get();
-
-        m_array_ptr2 = nullptr;
-        m_array_ptr2 = LeafPtr(new (&m_leaf_cache_storage2) LeafType(m_table.unchecked_ptr()->get_alloc()));
-        this->m_cluster->init_leaf(this->m_condition_column_key2, m_array_ptr2.get());
-        m_leaf_ptr2 = m_array_ptr2.get();
+        m_leaf_ptr1 =
```

Comment:
I guess there is no point in creating a new array for every cluster. The type would be the same every time.

Reply:
👍  updated